### PR TITLE
crossgen2: Add --imagebase option

### PR DESCRIPTION
**Pull request by t-mustafin**
--imagebase option set preferable ImageBase field to output PE-file

### CODE REVIEW

**Comment by t-mustafin** [comment]
cc @alpencolt @gbalykov

File: src/coreclr/tools/Common/Compiler/CompilationBuilder.cs [hunk 1]
```diff
@@ -25,6 +25,7 @@ public abstract partial class CompilationBuilder
         protected OptimizationMode _optimizationMode = OptimizationMode.None;
         protected int _parallelism = -1;
         protected bool _resilient;
+        protected ulong _imageBase;
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/Common/Compiler/CompilationBuilder.cs:28]: Could you move this to src\coreclr\tools\aot\ILCompiler.ReadyToRun\Compiler\ReadyToRunCodegenCompilationBuilder.cs?

This setting only makes sense for ReadyToRun and CompilationBuilder.cs is shared with the NativeAOT compiler.

Comment (t-mustafin) [reply src/coreclr/tools/Common/Compiler/CompilationBuilder.cs]: Yes, thanks.

**Comment by MichalStrehovsky** [comment]
Looks reasonable to me but adding @dotnet/crossgen-contrib for actual signoff.

I do wonder whether the change you actually need is to just respect the ImageBase in the source executable (and set an ImageBase when compiling the source assembly). If I'm reading this right, this will set the same ImageBase for all components of a composite build and they'll be all guaranteed to conflict.

**Review by gbalykov — APPROVED** [review]

**Comment by t-mustafin** [comment]
> I do wonder whether the change you actually need is to just respect the ImageBase in the source executable (and set an ImageBase when compiling the source assembly). If I'm reading this right, this will set the same ImageBase for all components of a composite build and they'll be all guaranteed to conflict.

Actually we do not use composite mode, but use pipelene mode (--single-file-compilation --out-near-input). For pipeline mod I will prepare another patch.

**Comment by MichalStrehovsky** [comment]
> Actually we do not use composite mode, but use pipelene mode (--single-file-compilation --out-near-input). For pipeline mod I will prepare another patch.

I assume the command line interface will get ugly for the pipeline mode. Is there anything preventing you from setting the ImageBase on each of the inputs _before_ running crossgen? The ImageBase of an assembly can be controlled at the csproj level of each assembly at the time of source build by using the [`<BaseAddress>`](https://docs.microsoft.com/en-us/visualstudio/msbuild/common-msbuild-project-properties?view=vs-2022) property. If we made crossgen2 respect the `BaseAddress` of the input module instead of overwriting it with a new value, it would probably end up cleaner.

**Comment by jkotas** [comment]
You can build a separate tool that goes over the universe of all assemblies and assigns them a base address based on some policy. crossgen can then pick the base address from each assembly in the pipeline mode, no command line option needed.

I understand that you are using fixed base addresses to save memory in Tizen specific builds via `FEATURE_ENABLE_NO_ADDRESS_SPACE_RANDOMIZATION`. Fixed base addresses suppress address space randomization that goes against our security roadmap, so it is not something we are likely to design a great experience for.

**Comment by trylek** [comment]
@t-mustafin - do you think that Jan's and Michal's suggestions are sufficient to address your needs or that there is still some value in making this change?

**Review by trylek — APPROVED** [review]
As I believe that @gbalykov's explanation is reasonable and this change basically reintroduces functionality that was present in the legacy native Crossgen1, I believe it should be safe to take. I'll leave the PR open until tomorrow to see whether there's additional pushback from @jkotas or @MichalStrehovsky; in its absence I'll just merge it in tomorrow morning. Thanks!

**Pull request merged**